### PR TITLE
[ENH] KCD and Bregman tests

### PR DESCRIPTION
**Pull request by adam2392**
Changes proposed in this pull request:
- Add KCD test and BregmanCDTest to the discrepancy/ submodule

To enable this and to simultaneously improve the readability and maintainability of the code:

- Refactor changes to enable code-sharing between kernel-based methods
- Refactored the delta kernel to be more readable and also added a unit-test. In addition, I made it in-line w/ the way that pairwise kernel functions should be, so it's compatible with sklearn API.
        - Specifically refactor the way that the kernels are computed, to make it more in-line with how sklearn does pairwise kernels. This leverages their `pairwise_kernels` function, which inherently is parallelized probably better than we can, and can leverage all possible kernels they have and more. Moreover, I added a pattern to demonstrate that we can augment the `PAIRWISE_KERNEL_FUNCTIONS` from sklearn whenever we need to add a kernel that is more efficient to compute using vectorized operations of numpy. This allows us to keep the `delta_kernel` as is.
- Speed up the kci unit tests. Before just the kci unit-tests took on average 10+ seconds to run. Now they run in less than 2 seconds with the same results to prevent regressions.

Now entire unit-test suite takes 8 seconds to run.

## Refactoring keeps functionality
None of the tests changed functionally. If anything, I lowered the number of samples used for most tests and they all still pass, so the refactoring does not introduce regressions.

## Before submitting

- [ ] I've read and followed all steps in the [Making a pull request](https://github.com/py-why/pywhy-stats/blob/main/CONTRIBUTING.md#making-a-pull-request)
    section of the `CONTRIBUTING` docs.
- [ ] I've updated or added any relevant docstrings following the syntax described in the
    [Writing docstrings](https://github.com/py-why/pywhy-stats/blob/main/CONTRIBUTING.md#writing-docstrings) section of the `CONTRIBUTING` docs.
- [ ] If this PR fixes a bug, I've added a test that will fail without my fix.
- [ ] If this PR adds a new feature, I've added tests that sufficiently cover my new functionality.
- [ ] I have added a changelog entry succintly describing the change in this PR in the [whats_new](https://github.com/py-why/pywhy-stats/blob/main/docs/whats_new/) relevant version document.

## After submitting

- [ ] All GitHub Actions jobs for my pull request have passed.

### CODE REVIEW

**Comment by adam2392** [comment]
We should have an example demonstrating how to leverage different kernels. We can ideally show how different kernels are useful too on a small simulated data.

**Comment by adam2392** [comment]
Hmm a bit weird, but I think the `preserve_random_seed` decorator messes up the sphinx auto-doc: https://output.circle-artifacts.com/output/job/9f13ba5a-fb34-49d5-be02-2af229ef8e88/artifacts/0/dev/generated/pywhy_stats.independence.kci.html#module-pywhy_stats.independence.kci

compared to https://output.circle-artifacts.com/output/job/9f13ba5a-fb34-49d5-be02-2af229ef8e88/artifacts/0/dev/generated/pywhy_stats.independence.fisherz.html#module-pywhy_stats.independence.fisherz

there is a missing `condind` and  `ind` function documentation.

**Comment by adam2392** [comment]
@bloebp just checking in to see if you had the opportunity to look at this?

File: pywhy_stats/discrepancy/base.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+        The null distribution of test statistics.
+    """
+    rng = np.random.default_rng(seed)
+
+    n_samps = X.shape[0]
+
+    # compute the test statistic on the conditionally permuted
+    # dataset, where each group label is resampled for each sample
+    # according to its propensity score
+    null_dist = Parallel(n_jobs=n_jobs)(
```
Comment (bloebp) [inline pywhy_stats/discrepancy/base.py:121]: Note that using the parallel jobs here would ignore the previously set random seed. This can be fixed doing something like here:
https://github.com/py-why/dowhy/blob/main/dowhy/gcm/independence_test/kernel.py#L101

The idea is to generate random seeds based on the current (seeded) random generator and provide these seeds to the parallel processes. In that way, the generated seeds are deterministic and, thus, the parallel processes.

File: pywhy_stats/discrepancy/bregman.py [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+
+def condind(
+    X: ArrayLike,
+    Y: ArrayLike,
+    group_ind: ArrayLike,
+    kernel: Optional[Callable[[ArrayLike], ArrayLike]] = None,
+    null_sample_size: int = 1000,
+    normalize_data: bool = False,
+    propensity_model=None,
+    propensity_weights=None,
```
Comment (bloebp) [inline pywhy_stats/discrepancy/bregman.py:34]: Something is off with the formatting here.

File: pywhy_stats/discrepancy/bregman.py [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+
+        P_{Z=1}(Y|X) = P_{Z=0}(Y|X)
+
+    Parameters
+    ----------
+    X : ArrayLike of shape (n_samples, n_features_x)
+        Data for variable X, which can be multidimensional.
+    Y : ArrayLike of shape (n_samples, n_features_y)
+        Data for variable Y, which can be multidimensional.
+    group_ind : ArrayLike of shape (n_samples,)
```
Comment (bloebp) [inline pywhy_stats/discrepancy/bregman.py:54]: This basically assigns each sample in `X` and `Y` to either group 0 or 1, right? The docstring can be more explicit in this regard.

I am also wondering it wouldn't be clearer if we just accept four inputs: `X_baseline, X_reference, Y_baseline, Y_reference`, where baseline is `Z=0` and reference `Z=1` (or something along the line). This would definitely inflate the number of input arguments, but also don't need the ground indices anymore.

File: pywhy_stats/discrepancy/kcd.py [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+        P_{Z=1}(Y|X) = P_{Z=0}(Y|X)
+
+    Parameters
+    ----------
+    X : ArrayLike of shape (n_samples, n_features_x)
+        Data for variable X, which can be multidimensional.
+    Y : ArrayLike of shape (n_samples, n_features_y)
+        Data for variable Y, which can be multidimensional.
+    group_ind : ArrayLike of shape (n_samples,)
+        Data for group indicator Z, which can be multidimensional.
```
Comment (bloebp) [inline pywhy_stats/discrepancy/kcd.py:60]: Same as in the other test.

File: pywhy_stats/discrepancy/kcd.py [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+    _preprocess_kernel_data,
+    compute_kernel,
+)
+
+from ..pvalue_result import PValueResult
+from .base import _compute_propensity_scores, _preprocess_propensity_data, compute_null
+
+
+# XXX: determine if we can do this with Y being optional.
+def condind(
```
Comment (bloebp) [inline pywhy_stats/discrepancy/kcd.py:30]: Here and in the bregman test, an "unconditional" version doesn't make sense right? That would basically mean we just check if two distributions are the same.

File: pywhy_stats/independence/kci.py [hunk 6]
```diff
@@ -1,15 +1,23 @@
-from functools import partial
```
Comment (bloebp) [inline pywhy_stats/independence/kci.py:1]: Is it possible to separate the changes in the existing kernel tests from adding the discrepancy tests? There are now several things/modifications we should check in a separate PR. If necessary, there can be a third PR then afterwards that combines the discrepancy tests with the changes in the kernel tests.

File: tests/discrepancy/test_cd.py [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+
+from pywhy_stats import bregman, kcd
+
+seed = 12345
+rng = np.random.default_rng(seed)
+# number of samples to use in generating test dataset; the lower the faster
+n_samples = 150
+
+
+def single_env_scm(n_samples=200, offset=0.0):
```
Comment (bloebp) [inline tests/discrepancy/test_cd.py:14]: If we only use this withing this test, add an `_` to the method.

File: tests/discrepancy/test_cd.py [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+import pandas as pd
+import pytest
+from sklearn.ensemble import RandomForestClassifier
+
+from pywhy_stats import bregman, kcd
+
+seed = 12345
+rng = np.random.default_rng(seed)
+# number of samples to use in generating test dataset; the lower the faster
+n_samples = 150
```
Comment (bloebp) [inline tests/discrepancy/test_cd.py:11]: Generally, rather use the `flaky` decorator with a 'reasonable' corresponding assert definition instead of a fixed random seed in unit tests. Happened too often that a bug was not captured because it just worked with a concrete seed. If we just generate random data, then we would capture it if it is consistently failing. I know the concern regarding the "random fails" or runtime in stochastic tests, but this is rarely a practical issue if we set the retries to like 3.

File: tests/test_kci.py [hunk 9]
```diff
@@ -1,19 +1,22 @@
...
+from sklearn.metrics.pairwise import rbf_kernel
 
 from pywhy_stats import kci
 from pywhy_stats.kernels import delta_kernel
 
 ####################################################
 # Unconditional independence tests - Continuous
 ####################################################
+seed = 12345
+rng = np.random.default_rng(seed)
```
Comment (bloebp) [inline tests/test_kci.py:12]: See the other comment regarding the seed

File: tests/test_kci.py [hunk 10]
```diff
@@ -1,19 +1,22 @@
...
 
 
 @flaky(max_runs=3)
 def test_given_continuous_independent_data_when_perform_kernel_based_test_then_not_reject():
-    x = np.random.randn(1000, 1)
-    y = np.exp(np.random.rand(1000, 1))
+    n_samples = 200
+    x = rng.standard_normal((200, 1))
```
Comment (bloebp) [inline tests/test_kci.py:18]: Any reason to change it to `rng` instead of using `numpy`? If it is to use the random generator, this is already covered by the flaky decorator of the test, i.e. no need to have a seed.

File: tests/test_kci.py [hunk 11]
```diff
@@ -27,9 +30,9 @@ def test_given_continuous_dependent_data_when_perform_kernel_based_test_then_rej
 
          X <- Z -> Y
     """
-    z = np.random.randn(1000, 1)
-    x = np.exp(z + np.random.rand(1000, 1))
-    y = np.exp(z + np.random.rand(1000, 1))
+    z = rng.standard_normal((1000, 1))
+    x = np.exp(z + rng.uniform(size=(1000, 1)))
+    y = np.exp(z + rng.uniform(size=(1000, 1)))
```
Comment (bloebp) [inline tests/test_kci.py:35]: Why uniform instead of Gaussian noise? It doesn't matter here, just curious to hear.

File: tests/discrepancy/test_cd.py [hunk 12]
```diff
@@ -0,0 +1,201 @@
...
+            X=df["x"],
+            Y=df["y"],
+            group_ind=df[group_col],
+            null_sample_size=null_sample_size,
+            n_jobs=-1,
+            **cd_kwargs,
+            random_seed=seed,
+        )
+        assert res.pvalue > alpha, f"Fails with {res.pvalue} not greater than {alpha}"
+    elif env_type == "multi":
```
Comment (bloebp) [inline tests/discrepancy/test_cd.py:171]: What about instead of the if-elifs, just separate them into two tests with `test_cd_simulation_single` and `test_cd_simulation_multi`.

Optimally, one would even have a separate test for each `assert` (but we don't need to over-engineer it).

Also, since you don't use the given-when-then pattern for the unit test names, add a brief description what is tested here.

**Review by bloebp — COMMENTED** [review]
Overall, it looks good! The main thing is: Can we separate the KCI test changes into a separate PR or do you need to have these modifications for the discrepancy tests?

File: tests/test_kci.py (hunk 10 above)
Comment (adam2392) [reply tests/test_kci.py:18]: It's quite difficult to debug tests if they're not deterministic, so the nice thing is if we break stuff and this test fails, we can keep running it to get the same answer and do some debugging. With the previous random seed, I spent a lot of time trying to figure out what was going on because the test didn't pass but the output was changing, so hard to isolate.

The flaky thing will still work because the CIs inherently test on different distros/OS/etc., so this will still have the same features, but just locally you can reproduce failures to the exact numerical precision.

File: pywhy_stats/discrepancy/base.py [hunk 13]
```diff
@@ -0,0 +1,127 @@
+from typing import Callable, Optional
```
[the remark this answers is not in the document]
Comment (adam2392) [reply pywhy_stats/discrepancy/base.py:1]: These should be able to be used for all discrepancy tests since testing `P(y|x) = P'(y|x)` usually requires some type of "matching" done on the covariates, which in this case is done via the propensity scores.

File: doc/conf.py [hunk 14]
```diff
@@ -146,9 +146,8 @@
     "PValueResult": "pywhy_stats.pvalue_result.PValueResult",
     # numpy
     "NDArray": "numpy.ndarray",
-    # "ArrayLike": "numpy.typing.ArrayLike",
     "ArrayLike": ":term:`array_like`",
-    "fisherz": "pywhy_stats.fisherz",
+    # "fisherz": "pywhy_stats.independence.fisherz",
```
[the remark this answers is not in the document]
Comment (adam2392) [reply doc/conf.py:150]: Done.

File: doc/api.rst [hunk 15]
```diff
@@ -57,10 +57,24 @@ contains the p-value and the test statistic and optionally additional informatio
...
+.. currentmodule:: pywhy_stats.independence
 .. autosummary::
    :toctree: generated/
-
+   
    fisherz
    kci
 
+
+(Conditional) Discrepancy Testing
```
[the remark this answers is not in the document]
Comment (adam2392) [reply doc/api.rst:68]: Perhaps Conditional K-Sample Testing?

I use discrepancy because I saw it in one paper, but I think (conditional) k-sample is more correct. I guess in our cases, we are only doing conditional 2-sample testing, since it's P_1(y|x) =? P_2(y|x)

[the remark this answers is not in the document]
Comment (adam2392) [reply doc/api.rst:68]: I changed it

File: pywhy_stats/discrepancy/base.py [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+    random_state: Optional[int] = None,
+    **kwargs,
+):
+    if propensity_model is None:
+        K: ArrayLike = kwargs.get("K")
+
+        # compute a default penalty term if using a kernel matrix
+        if K.shape[0] == K.shape[1]:
+            propensity_penalty_ = _default_regularization(K)
+            C = 1 / (2 * propensity_penalty_)
```
[the remark this answers is not in the document]
Comment (adam2392) [reply pywhy_stats/discrepancy/base.py:54]: This just allows us to magically infer a default regularization for LogisticRegression for propensity score estimation. Since LogisticRegression is sort of the de-facto default in literature, this seems like a reasonable default. Alternatively, `propensity_model` can be specified by the user fully, which can be for example a fully specified RandomForestClassifier.

[the remark this answers is not in the document]
Comment (adam2392) [reply pywhy_stats/discrepancy/base.py:54]: I added some inline comments too

File: pywhy_stats/discrepancy/bregman.py (hunk 2 above)
Comment (adam2392) [reply pywhy_stats/discrepancy/bregman.py:34]: Fixed.

File: pywhy_stats/discrepancy/bregman.py (hunk 3 above)
Comment (adam2392) [reply pywhy_stats/discrepancy/bregman.py:54]: I use `group_ind` because this would allow us to generalize these tests to `k-sample` in the future. I know some groups are working on this. I suppose we could for now do the 4 inputs, but later on if we generalize to k-sample, then we need to refactor to this design.

> This basically assigns each sample in `X` and `Y` to either group 0 or 1, right? The docstring can be more explicit in this regard.

Yep. Fixed the docstring.

File: pywhy_stats/discrepancy/kcd.py (hunk 4 above)
Comment (adam2392) [reply pywhy_stats/discrepancy/kcd.py:60]: Done.

File: tests/discrepancy/test_cd.py (hunk 7 above)
Comment (adam2392) [reply tests/discrepancy/test_cd.py:14]: Done

File: tests/test_kci.py (hunk 11 above)
Comment (adam2392) [reply tests/test_kci.py:35]: `rand` is `uniform` essentially. I think this is just an example I followed off the literature. I think it also helps limit the amount of variability since there's a nonlinear transformation going on.

I just tried with `standard_normal` and it works too actually. Preference on keeping either one?

File: pywhy_stats/discrepancy/base.py (hunk 1 above)
Comment (adam2392) [reply pywhy_stats/discrepancy/base.py:121]: I think this one is fine because it uses the `rng.binomial` is not passed into the inner function. I added a unit-test though just for completeness tho.

File: tests/discrepancy/test_cd.py (hunk 12 above)
Comment (adam2392) [reply tests/discrepancy/test_cd.py]: Refactored and also renamed tests.

**Comment by adam2392** [comment]
I addressed most of your comments and will refactor now to a 2-stage PR. I'll be able to do this sometime later on.

File: pywhy_stats/kernel_utils.py [hunk 17]
```diff
@@ -0,0 +1,281 @@
...
+
+    # compute the potentially pairwise kernel matrix
+    # If the number of arguments is just one, then we bypass the pairwise kernel
+    # optimized computation via sklearn and opt to use the metric function directly
+    if callable(metric) and len(inspect.getfullargspec(metric).args) == 1:
+        kernel = metric(X)
+    else:
+        kernel = pairwise_kernels(
+            X, Y=Y, metric=metric, n_jobs=n_jobs, filter_params=False, **kernel_params
+        )
```
Comment (bloebp) [inline pywhy_stats/kernel_utils.py:142]: Can we move the change of allowing strings as kernels instead of only callable into a separate PR? Would like to take a separate look at this, which is not mixed together with other changes. For this PR, maybe stick to passing a callable (e.g., a kernel from scikit-learn via `kernel_X = partial(pairwise_kernels, metric=metric, n_jobs=n_jobs, filter_params=False, other_params...)` which should be equivalent).

File: pywhy_stats/kernel_utils.py [hunk 18]
```diff
@@ -0,0 +1,281 @@
...
+
+            # compute the sample centered correntropy
+            corren = (1.0 / n_samples) * np.trace(K) - bias
+
+            corren_arr[idx, jdx] = corren_arr[jdx, idx] = corren
+    return corren_arr
+
+
+def von_neumann_divergence(A: ArrayLike, B: ArrayLike) -> float:
+    """Compute Von Neumann divergence between two PSD matrices.
```
Comment (bloebp) [inline pywhy_stats/kernel_utils.py:205]: What does PSD stand for? Maybe write it out.

File: pywhy_stats/discrepancy/kcd.py [hunk 19]
```diff
@@ -0,0 +1,262 @@
...
+        null_sample_size=null_sample_size,
+        normalize_data=normalize_data,
+        centered=centered,
+        n_jobs=n_jobs,
+        random_seed=random_seed,
+    )
+    return PValueResult(pvalue=pvalue, statistic=test_statistic)
+
+
+def _kernel_test(
```
[the remark this answers is not in the document]
Comment (adam2392) [reply pywhy_stats/discrepancy/kcd.py:120]: This is different from the `_kernel_test` inside `kci.py`. It is just named the same for consistency. What does overlap though are the functions used for computing kernels, (fast) centering them, and preprocessing/validating the input data.

[the remark this answers is not in the document]
Comment (bloebp) [reply pywhy_stats/discrepancy/kcd.py:120]: Ah ok got it, then all good!

File: pywhy_stats/discrepancy/kcd.py [hunk 20]
```diff
@@ -0,0 +1,262 @@
...
+    X: ArrayLike,
+    group_ind: ArrayLike,
+    Y: ArrayLike,
+    kernel_X: Optional[Callable[[np.ndarray], np.ndarray]],
+    kernel_Y: Optional[Callable[[np.ndarray], np.ndarray]],
+    propensity_weights: Optional[ArrayLike],
+    propensity_model: Optional[BaseEstimator],
```
[the remark this answers is not in the document]
Comment (bloebp) [reply pywhy_stats/discrepancy/kcd.py:127]: My confusion was that I thought this method is shared with the KCI test. But it is not, all good then!

File: tests/test_kci.py (hunk 11 above)
Comment (adam2392) [reply tests/test_kci.py:35]: `rand` is uniform :p: https://numpy.org/doc/stable/reference/random/generated/numpy.random.rand.html

Okay yeah it seems robust, so I'll just leave it as it was previously.

File: pywhy_stats/__init__.py [hunk 21]
```diff
@@ -1,3 +1,5 @@
-from . import fisherz, kci
+from . import discrepancy, independence
```
[the remark this answers is not in the document]
Comment (adam2392) [reply pywhy_stats/__init__.py:1]: Okay perhaps, `conditional_ksample`?

File: tests/test_kci.py (hunk 11 above)
Comment (bloebp) [reply tests/test_kci.py:35]: Oh there is `rand` and `randn` haha
Seems that was even used in a mix before.

File: pywhy_stats/__init__.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (bloebp) [reply pywhy_stats/__init__.py:1]: Sounds good!

File: pywhy_stats/kernel_utils.py [hunk 22]
```diff
@@ -0,0 +1,281 @@
...
+    if X.ndim == 1:
+        X = X.reshape(-1, 1)
+    if Y.ndim == 1:
+        Y = Y.reshape(-1, 1)
+    if Z is not None and Z.ndim == 1:
+        Z = Z.reshape(-1, 1)
+
+    # handle strings as categorical data automatically
+    if X.dtype.type in (np.str_, np.object_, np.unicode_):
+        enc = LabelEncoder()
```
[the remark this answers is not in the document]
Comment (adam2392) [reply pywhy_stats/kernel_utils.py:262]: Do you have a strong preference for leaving the `np.str_` support in? Idr why, but I think I was running into a bunch of errors on the string unit-test; I'll need to look at how to revert this. In general, I find working internally assuming `ArrayLike[numbers.Real]` type easier to maintain, so we don' have separate code paths if something is a string. 

I suppose the main thing is being able to auto-assume something is categorical, with the caveat that the strings are correct. However, is the benefit worth it? 

Let me investigate in the separate PR to see if I can demonstrate what I mean.

[the remark this answers is not in the document]
Comment (bloebp) [reply pywhy_stats/kernel_utils.py:262]: In DoWhy-GCM, we have the convention that categorical data are strings. So, if I want to use this method with string data and do not define a particular kernel (which majority of the users won't), it would not work. The easiest convention is to define all string values as categorical. I actually simply checked before if the first element in a column value is a `str`, then we don't need all this `np.str_` types. 

An alternative (but technically less accurate) is to introduce something like `encode_strings: bool = False`, default to RBF kernels and then throw an error with string data if `encode_strings` is set to False. But again, applying the delta kernel would be more accurate.

File: pywhy_stats/kernel_utils.py (hunk 17 above)
Comment (adam2392) [reply pywhy_stats/kernel_utils.py:142]: Sure thing!

Note: The change here is made to allow maximum flexibility on the user, while not having to write our own pairwise kernel code for custom kernels.

tldr: if we adhere to the scikit-learn pattern, we have less code and buys some helper functions for free, but we need to restrict what type of kernel functions are allowed. 

The issue here is that custom kernels are slow, but that is always the case. However, in order to provide some consistency to make maintenance and code simpler, we probably want to be consistent with how we define a kernel function. That is why I added the extra unused parameters to the delta kernel. Having the `def kernel_func(X, Y=None, **kwargs)` pattern, allows us to re-use the `pairwise_kernels` function.

But it also buys us the usages of strings for free, since they accept those.

Alternatively, if we want to not allow strings, and use the `partial(func(...))` pattern, then we have to write our own parallelized pairwise_kernel function.

Limiting the scope of what we support is probably a good idea, so users can't just feed in some arbitrary lambda functions that we can't easily error check. 

Idk how to handle the `custom_kernel` you define that combines a delta and rbf kernel, so I added this hack for now. However, this raises a question for me: is a custom kernel like that even valid?

Comment (bloebp) [reply pywhy_stats/kernel_utils.py:142]: I think most users can simply use a kernel from scikit. We can definitely think about allowing strings as well, just want to have this review separate.

> However, this raises a question for me: is a custom kernel like that even valid

It is, a convolution of different valid kernels is also a valid kernel and, practically, you can achieve this by pointwise multiplication of the resulting kernel matrices.
